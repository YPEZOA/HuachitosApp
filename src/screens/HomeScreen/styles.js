import { StyleSheet } from 'react-native'
import colors from '../../UI/colors'

export const homeScreen = StyleSheet.create({
  containerMain: {
    flex: 1,
    paddingTop: 50,
    paddingHorizontal: 10,
  },
  scrollContainer: {
    marginTop: 30,
  },
  inputContainer: {
    marginTop: 20,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    padding: 5,
    backgroundColor: '#dcdcdc',
    borderRadius: 100,
  },
  inputIcon: {
    padding: 10,
    borderRadius: 100,
    backgroundColor: colors.white,
  },
  headerLogoContent: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  headerText: {
    fontSize: 28,
    fontFamily: 'Quicksand-Bold',
    color: colors.primary,
  },
  title: {
    marginTop: 20,
    fontSize: 25,
    fontWeight: 'bold',
    color: '#5e5e5e',
  },
})
