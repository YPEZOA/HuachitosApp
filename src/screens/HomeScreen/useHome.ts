import { useEffect, useState } from 'react'
import { getAllAnimals, getAnimalsByFilters, getAnimalsByType } from '../../services/animals'
import { isFetching } from '../../utils/helpers/loader'

const initialState = {
  state: '',
  region: { id: null, region: '', comunas: [] },
  comuna: null,
}

const useHome = () => {
  const [allAnimals, setAllAnimals] = useState([])
  const [animalTypeSelected, setAnimalTypeSelected] = useState('')
  const [fetching, setFetching] = useState(false)
  const [showFiltersModal, setShowFiltersModal] = useState(false)
  const [filterParameters, setFilterParameters] = useState(initialState)
  const [animalsFiltered, setAnimalsFiltered] = useState([])
  const [filterSubmited, setFilterSubmited] = useState(false)
  const [fetchWithoutData, setFetchWithoutData] = useState(false)

  const fetchWrap = isFetching(setFetching)

  useEffect(() => {
    async function getAnimals() {
      setFetching(true)
      const animals = await getAllAnimals()
      dataIsVisible(animals)
      setAllAnimals(animals.data)
      setFetching(false)
    }
    getAnimals()
  }, [])

  const dataIsVisible = ({ data }: any) => {
    return !data?.length ? setFetchWithoutData(true) : setFetchWithoutData(false)
  }

  const resetFilter = () => {
    setFilterSubmited(false)
    setFilterParameters(initialState)
    setAnimalsFiltered([])
    setFilterSubmited(false)
  }

  return {
    states: {
      fetchWithoutData,
      filterSubmited,
      allAnimals,
      animalTypeSelected,
      fetching,
      showFiltersModal,
      filterParameters,
      animalsFiltered,
      get getAvailableCommunes() {
        return filterParameters.region.comunas
      },
      get anyFilterSelected() {
        return filterParameters !== initialState
      },
      get filterResultLength() {
        return animalsFiltered?.length > 0
      },
    },
    setters: {
      setFilterSubmited,
      setAnimalTypeSelected,
      setShowFiltersModal,
      setFilterParameters,
      setAnimalsFiltered,
    },
    actions: {
      resetFilter,
      filterAnimalsByType: fetchWrap(async (type: string) => {
        if (!type.length) return
        const animalsFilter = await getAnimalsByType(type)
        dataIsVisible(animalsFilter)
        setAnimalsFiltered(animalsFilter.data)
      }),
      getAnimalsByFiltered: fetchWrap(async () => {
        const filterPayload = {
          ...filterParameters,
          region: filterParameters.region.id,
        }
        const { comuna, state, region } = filterPayload

        if (filterPayload.comuna !== null) {
          const animalsByComuna = await getAnimalsByFilters(
            `comuna/${comuna}${state.length ? `/estado/${state}` : ''}`
          )
          dataIsVisible(animalsByComuna)
          setAnimalsFiltered(animalsByComuna.data)
          setAnimalTypeSelected('')
          setFilterParameters(initialState)
        } else {
          const regionSelected = filterPayload.region !== null
          const stateSelected = filterPayload.state !== ''
          const animalsFilteredBySelection = await getAnimalsByFilters(
            `${
              regionSelected && stateSelected
                ? `region/${region}/estado/${state}`
                : regionSelected
                ? `region/${region}`
                : `estado/${state}`
            }`
          )
          setAnimalTypeSelected('')
          setAnimalsFiltered(animalsFilteredBySelection.data)
          dataIsVisible(animalsFilteredBySelection)
          setFilterParameters(initialState)
        }
      }),
    },
  }
}

export default useHome
