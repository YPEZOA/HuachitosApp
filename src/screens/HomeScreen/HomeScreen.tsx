import React, { useEffect } from 'react'
import { useNavigation } from '@react-navigation/native'
import { Text, View } from 'react-native'
import { homeScreen as St } from './styles'
import { getAllAnimals } from '../../utils/helpers/http'
import Animals from './components/Animals/Animals'
import Icon from 'react-native-vector-icons/FontAwesome'
import colors from '../../UI/colors'

const HomeScreen = () => {
  useEffect(() => {
    getAllAnimals()
  }, [])

  const navigation = useNavigation()

  return (
    <View style={St.containerMain}>
      <View style={St.headerLogoContent}>
        <Text style={St.headerText}>
          Huachit
          <Icon name="paw" size={20} />s
        </Text>
        <Text style={{ color: colors.primary, fontSize: 10 }}>
          Encuentra y da familia
        </Text>
      </View>
      <Animals />
    </View>
  )
}

export default HomeScreen
