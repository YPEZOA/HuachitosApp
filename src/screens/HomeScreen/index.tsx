import React from 'react'
import { TouchableOpacity, View, StatusBar, Pressable, Text } from 'react-native'
import { homeScreen as St } from './styles'
import { HomeContext } from './context'
import useHome from './useHome'
import Animals from './components/Animals/Animals'
import IconF from 'react-native-vector-icons/Ionicons'
import Filters from './components/Filter/Filters'
import FiltersModal from './components/FiltersModal/FiltersModal'
import LightLogoText from '../../UI/Icons/LogoWithTagLight'
import FilterResult from './components/FilterResult/FilterResult'
import { useNavigation } from '@react-navigation/native'
import colors from '../../UI/colors'

const HomeScreen = () => {
  const { states, actions, setters } = useHome()
  const navigation = useNavigation()

  return (
    <HomeContext.Provider value={{ states, actions, setters }}>
      <StatusBar backgroundColor={colors.primary} />
      {/* <Pressable onPress={() =>  */}
      {/*   <Text>FAVORITES</Text> */}
      {/* </Pressable> */}
      <View style={St.containerMain}>
        {/* Header */}
        <View style={St.headerContainer}>
          <LightLogoText width={220} height={70} />
          <View style={St.headerButtonsContainer}>
            <TouchableOpacity
              onPress={() => navigation.navigate('AnimalsLikes')}
              style={St.closeIcon}
            >
              <IconF name="heart" size={30} color="#ee6352" />
            </TouchableOpacity>
            <TouchableOpacity
              onPress={() => setters.setShowFiltersModal(true)}
              style={St.closeIcon}
            >
              <IconF name="filter" size={30} color={colors.secondary} />
            </TouchableOpacity>
          </View>
        </View>
        {/* Animals list */}
        <View style={St.containerContent}>
          <Filters />
          {states.filterSubmited ? <FilterResult /> : <Animals />}
        </View>
      </View>
      {/* Filters Modal */}
      <FiltersModal />
    </HomeContext.Provider>
  )
}

export default HomeScreen
