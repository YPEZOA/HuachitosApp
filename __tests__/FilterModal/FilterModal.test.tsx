import React from 'react'
import { fireEvent, render, renderHook, waitFor } from '@testing-library/react-native'
import HomeScreen from '../../src/screens/HomeScreen'
import useHome from '../../src/screens/HomeScreen/useHome'
import { act } from 'react-test-renderer'

describe('Test <FilterModal/>', () => {
  test('Should be component visible', () => {
    const { toJSON, getByTestId } = render(<HomeScreen />)
    const openFilterBtn = getByTestId('open-filters')
    fireEvent.press(openFilterBtn)

    const modalInstance = getByTestId('filters-modal')
    expect(modalInstance).toBeDefined()
    expect(toJSON).toMatchSnapshot()
  })

  test('Should show disabled button to apply filters', () => {
    const { getByTestId } = render(<HomeScreen />)
    const openFilterBtn = getByTestId('open-filters')
    fireEvent.press(openFilterBtn)

    const applyFiltersBtn = getByTestId('apply-filters')
    expect(applyFiltersBtn).toBeDefined()
    expect(applyFiltersBtn.props.accessibilityState.disabled).toBeTruthy()
  })

  test('Should be show enabled button', async () => {
    const { result } = renderHook(() => useHome())
    const { getByTestId, getAllByTestId } = render(<HomeScreen />)
    const openFilterBtn = getByTestId('open-filters')
    fireEvent.press(openFilterBtn)

    const animalStateButtons = getAllByTestId('animal-state')
    expect(animalStateButtons.length).toBe(3)

    fireEvent.press(animalStateButtons[0])
    const { states, setters } = result.current

    await waitFor(() => {
      setters.setFilterParameters({ ...states.filterParameters, state: 'adopcion' })
      const func = jest.fn(() => states.anyFilterSelected())
      func()
    })
  })
})
