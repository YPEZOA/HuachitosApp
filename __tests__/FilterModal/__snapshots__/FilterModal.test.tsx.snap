// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <FilterModal/> Should be component visible 1`] = `[Function]`;
